api/v1_projects.inc

```diff
@@ -271,6 +271,93 @@ function render_project_page_json($row)
     ];
 }
 
+// POST v1/projects/:projectid/pages/checkout
+// checkout a new page to proofread
+function api_v1_project_checkout($method, $data, $query_params)
+{
+    global $pguser;
+
+    $project = $data[":projectid"];
+    $required_roundid = $query_params["roundid"];
+    validate_user_can_work_in_project($project, $required_roundid);
+
+    try {
+        $lpage = get_available_page($project->projectid, $project->state, $pguser, $err);
+    } catch (ProjectPageException $exception) {
+        throw new BadRequest($exception->getMessage());
+    }
+
+    if (is_null($lpage)) {
+        throw new BadRequest($err);
+    }
+    return ["pagename" => $lpage->imagefile, "page_state" => $lpage->page_state];
+}
+
+// PUT projects/:projectid/pages/:pagename/rounds/:roundid
+// the PUTs all can change the page text or page state.
+// they are idempotent in the sense that if the identical request is issued
+// again either the the state and text will not be changed or the request will
+// be rejected because the page state has changed so there will be no change
+// to the server. Logging side-effects excluded.
+function api_v1_project_page_round($method, $data, $query_params)
+{
+    global $pguser;
+
+    $project = $data[":projectid"];
+    $round = $data[":roundid"];
+    validate_user_can_work_in_project($project, $round->id);
+    $request_body = api_get_request_body();
+    try {
+        $page_state = $request_body->page_state;
+        $lpage = get_indicated_LPage($project->projectid, $project->state, $data[":pagename"], $page_state, 0);
+        $action = $request_body->action;
+
+        switch ($action) {
+        case "return":
+            $lpage->returnToRound($pguser);
+            return [];
+            break;
+        case "save_temp":
+            $lpage->saveAsInProgress($request_body->text, $pguser);
```

Comment:
We don't need to return the page text in this case unless we allow that it could be different because the js validator is not identical to the php one. But we not concerned about this in the 'save as done' case.

Reply:
You're making the incorrect assumption that the only user of this API could be the UI you are building that uses the JS validator. Anyone can use this API. We may (or may not) want to return the page text in this call, but the schema should be the same for every call.

Reply:
Good point.